kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/Metadata.kt

```diff
@@ -0,0 +1,350 @@
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      flexibleTypeUpperBound = WildcardTypeName.subtypeOf(it)
+    }
+  }
+
+  override fun visitOuterType(flags: Flags): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      outerType = WildcardTypeName.supertypeOf(it)
+    }
+  }
+
+  override fun visitStarProjection() {
+    argumentList.add(WildcardTypeName.STAR)
+  }
+
+  override fun visitTypeAlias(name: kotlinx.metadata.ClassName) {
+    if (!useTypeAlias) {
+      return
+    }
+    typeAliasName = name
+  }
+
+  override fun visitTypeParameter(id: Int) {
+    typeParameter = getTypeParameter?.invoke(id) ?: throw IllegalStateException(
+        "Visiting TypeParameter when there are no type parameters!")
+  }
+
+  override fun visitEnd() {
+    var finalType = flexibleTypeUpperBound ?: outerType ?: typeParameter
+    if (finalType == null) {
+      if (useTypeAlias) {
+        finalType = typeAliasType ?: typeAliasName?.let { ClassName.bestGuess(it.replace("/", ".")) }
+      }
+      if (finalType == null) {
+        finalType = className?.let { ClassName.bestGuess(it.replace("/", ".")) }
+            ?: throw IllegalStateException("No valid typename found!")
+        if (argumentList.isNotEmpty()) {
+          finalType = (finalType as ClassName).parameterizedBy(*argumentList.toTypedArray())
+        }
+      }
+    }
+
+    finalType = finalType.asNullableIf(nullable)
+    receiver(finalType)
+  }
+}
+
+internal fun KotlinClassMetadata.Class.readClassData(): ClassData {
+  @Suppress("RedundantExplicitType")
+  var classFlags: Flags = 0
+  lateinit var className: String
+  var constructorData: ConstructorData? = null
+  var companionObjectName: String? = null
+  val typeParameters = LinkedHashMap<Int, TypeVariableName>()
+  val typeParamResolver = { id: Int -> typeParameters[id]!! }
+  val superTypes = mutableListOf<TypeName>()
+  val properties = mutableListOf<PropertyData>()
+  accept(object : KmClassVisitor() {
+    override fun visit(flags: Flags, name: kotlinx.metadata.ClassName) {
+      super.visit(flags, name)
+      className = name
+      classFlags = flags
+    }
+
+    override fun visitTypeParameter(flags: Flags,
+        name: String,
+        id: Int,
+        variance: KmVariance): KmTypeParameterVisitor? {
+      return object : KmTypeParameterVisitor() {
+        val upperBoundList = mutableListOf<TypeName>()
+        override fun visitUpperBound(flags: Flags): KmTypeVisitor? {
+          return TypeNameKmTypeVisitor(flags) {
+            upperBoundList += it
+          }
+        }
+
+        override fun visitEnd() {
+          val finalVariance = variance.asKModifier().let {
+            if (it == KModifier.OUT) {
+              // We don't redeclare out variance here
+              null
+            } else {
+              it
+            }
+          }
+          val typeVariableName = if (upperBoundList.isEmpty()) {
```

Comment:
this doesn’t get populated with Any, does it? Would be nice if it did

Reply:
as in to cover the `NULLABLE_ANY` default case?